<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Стиль-заголовков" data-toc-modified-id="Стиль-заголовков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Стиль заголовков</a></span></li><li><span><a href="#Аномальные-значения" data-toc-modified-id="Аномальные-значения-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Аномальные значения</a></span></li><li><span><a href="#Пропуски-значений" data-toc-modified-id="Пропуски-значений-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Пропуски значений</a></span></li><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Определение-факторов,-влияющих-на-вероятность-просрочки-кредита" data-toc-modified-id="Определение-факторов,-влияющих-на-вероятность-просрочки-кредита-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Определение факторов, влияющих на вероятность просрочки кредита</a></span><ul class="toc-item"><li><span><a href="#Влияние-количества-детей-на-возврат-кредита" data-toc-modified-id="Влияние-количества-детей-на-возврат-кредита-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Влияние количества детей на возврат кредита</a></span></li><li><span><a href="#Влияние-семейного-положения-на-возврат-кредита" data-toc-modified-id="Влияние-семейного-положения-на-возврат-кредита-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Влияние семейного положения на возврат кредита</a></span></li><li><span><a href="#Влияние-уровня-дохода-на-возврат-кредита" data-toc-modified-id="Влияние-уровня-дохода-на-возврат-кредита-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Влияние уровня дохода на возврат кредита</a></span></li><li><span><a href="#Влияние-цели-кредита-на-его-возврат-в-срок" data-toc-modified-id="Влияние-цели-кредита-на-его-возврат-в-срок-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Влияние цели кредита на его возврат в срок</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></div>

# Исследование надёжности заёмщиков

**Цель исследования** — определить факторы, которые влияют на вероятность просрочки кредита.

**Ход исследования:**

Данные о поведении пользователей мы получим из файла `real_estate_data.csv`. О качестве данных ничего не известно. Поэтому перед определением взаимосвязей понадобится обзор данных.

Затем, на этапе предобработки, мы проверим данные на ошибки, оценим их влияние на исследование и поищем возможность исправить самые критичные ошибки данных.

На третьем этапе мы должны найти ответы на следующие вопросы заказчика:
* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Определение факторов, влияющих на вероятность просрочки кредита.

## Обзор данных

Составим первое представление о данных.




Основной инструмент аналитика — `pandas`. Импортируем эту библиотеку.

In [1]:
# импорт библиотеки pandas
import pandas as pd

Прочитаем файл `data.csv` и сохраним его в переменной `df`:

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('data.csv')

Выведем на экран первые десять строк таблицы:

In [3]:
# получение первых 10 строк таблицы df
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим общую информацию о таблице:

In [4]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Итак, в таблице 12 столбцов: в семи столбцах — числовые данные (в пяти — целочисленные, в двух — вещественные), еще в пяти столбцах — данные типа `object`. При этом вещественные данные удобнее трансформировать в целочисленные.

Данные сгруппированы по следующим столбцам:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;  
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения.
* `gender` — пол клиента;
* `income_type` — тип занятости;  
* `debt` — наличие задолженности по возврату кредита;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита;

В названиях колонок есть два нарушения стиля:
1. Непонятное название колонки с возрастом — `dob_years` вместо `age_years`.
2. Некорректное название колонки с типом занятости — `income_type` вместо `employment_type`.

Нелогичными выглядят также идентификаторы уровня образования и семейного положения: логичнее присвоить среднему образованию идентификатор `0`, а высшему образованию — `1`. Аналогично гражданскому браку логичнее присвоить `0`, а зарегистрированным семейным отношения — идентификатор `1`. Однако это не имеет критического значения.

В колонке `education` есть значения с разным регистром. Это необходимо будет исправить.

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения. Их доля составляет 10%, что существенно. Проверим, насколько это критично для исследования, с чем это связано (пропуски присутствуют только в двух столбцах — в столбце со стажем и с величиной дохода) и можно ли исправить.

Теперь оценим числовые переменные:

In [5]:
# оценка числовых переменных
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Видны следующие ошибки в числовых данных:

* `children` — есть отрицательные значения (`-1`), есть значение в `20` детей;
* `days_employed` — также есть отрицательные значения, и их много, есть и значение в 1100 лет стажа (`401.755` дней);
* `dob_years` — имеются строки с возрастом клиента `0`, что невозможно;
* `education_id` — имеются строки со значением `4`, что требует дополнительного анализа;
* `family_status_id	` — имеются строки со значением `4`, что, возможно, неверно и также требует дополнительного анализа.

**Выводы**

В каждой строке таблицы — данные об одним клиенте. Часть колонок описывает его в качественных категориях: количество детей, уровень образования, семейное положение, пол, тип занятости, наличие или отсутствие в прошлом просрочки по кредитам, цель получения кредита. Остальные категории — количественные: трудовой стаж, возраст, ежемесячный доход. 

Предварительно можно утверждать, что данных достаточно для проверки гипотез — имеется более 19 тыс. полностью заполненных строк. Но доля пропусков высока (10%), в названиях колонок присутствуют расхождения с хорошим стилем, а сами значения требуют обработки и исправления.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных
Исправим заголовки столбцов и проведём различные трансформации с самими данными.

### Стиль заголовков
Выведем на экран названия столбцов:

In [6]:
# перечень названий столбцов таблицы df
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Приведём названия столбцов в соответствие с хорошим стилем:
* заменим `dob` на `age`,
* заменим `income` на `employment`.

Для этого переименуем колонки так:
* `'dob_years'` → `'age_years'`;
* `'income_type'` → `'employment_type'`.

In [7]:
# переименование столбцов
df = df.rename(columns = {'dob_years':'age_years', 'income_type':'employment_type'})

Проверим результат. Для этого ещё раз выведем на экран названия столбцов:

In [8]:
# проверка результатов - перечень названий столбцов
df.columns

Index(['children', 'days_employed', 'age_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'employment_type',
       'debt', 'total_income', 'purpose'],
      dtype='object')

### Аномальные значения
Проанализируем аномальные значения. 
Начнём со столбца `children`. Выведем на экран все уникальные значения столбца и их частоту:

In [9]:
# вывод уникальных значений столбца 'children' и их частоты
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Во-первых, число детей не может быть отрицательным. Во-вторых, резкий переход от значений `5` к `20` и частота последних заставляет предположить банальную ошибку ввода данных, когда вместо `2` ввели `20` (неявный дубликат).

Заменим значения `-1` на `1` сейчас, а значения `20` - на `2` в параграфе **Дубликаты**.

In [10]:
# замена аномальных значений столбца 'children'
df['children'] = abs(df['children'])

Теперь обратимся к столбцу `days_employed` и проделаем аналогичную трансформацию отрицательных значений трудового стажа:

In [11]:
# замена аномальных значений столбца 'days_employed'
df['days_employed'] = abs(df['days_employed'])

Посмотрим теперь на максимальные и минимальные значения столбца.

In [12]:
# максимальные и минимальные значения столбца 'days_employed'
df.sort_values(by='days_employed', ascending = False)

children  days_employed  age_years education  education_id  \
6954          0  401755.400475         56   среднее             1   
10006         0  401715.811749         69    высшее             0   
7664          1  401675.093434         61   среднее             1   
2156          0  401674.466633         60   среднее             1   
7794          0  401663.850046         61   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender employment_type  debt  \
6954          вдовец / вдова                 2      F       пенсионер     0   
10006  Не женат / не замужем                 4      F       пенсионер     0   
7664         женат / замужем                 0      F       пенсионер     0   
2156         женат / замужем                 0      M       пенсионер     0   
7794        гражданский брак                 1      F       пенсионер     0   
...                      ...               ...    ...             ...   ...   
21489        женат / замужем                 0      M       компаньон     0   
21495       гражданский брак                 1      F       сотрудник     0   
21497        женат / замужем                 0      F       компаньон     0   
21502        женат / замужем                 0      F       сотрудник     0   
21510        женат / замужем                 0      F       сотрудник     0   

        total_income                           purpose  
6954   176278.441171                      ремонт жилью  
10006   57390.256908             получение образования  
7664   126214.519212                 операции с жильем  
2156   325395.724541                        автомобили  
7794    48286.441362                           свадьба  
...              ...                               ...  
21489            NaN              сделка с автомобилем  
21495            NaN                           свадьба  
21497            NaN        строительство недвижимости  
21502            NaN  строительство жилой недвижимости  
21510            NaN           приобретение автомобиля  

[21525 rows x 12 columns]

Значения трудового стажа менее месяца для людей старше 30 лет маловероятны, но проверить их не представляется возможным. Предположим, что они истинные.

Однако не может быть стажа в 1100 лет. В нашей таблице максимальный возраст клиента - 75 лет. Минимальный возраст для заключения трудового договора - 14 лет. Поэтому максимально возможный стаж для нашей таблицы равен 22 265 дням = 365 * (75 - 14). Возможно, банковский менеджер вместо 401755 дней стажа хотел указать 40,1755 лет стажа, но этого мы уже не узнаем.

Заменим с помощью функции значения стажа выше 22 265 дней на макcимально возможные в каждом случае, с учетом возраста каждого клиента.

Однако перед этим устраним аномалии в столбце `age_years`.

Посмотрим на уникальные значения столбца и их частоту.

In [13]:
# вывод уникальных значений столбца 'age_years' и их частоты
df['age_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: age_years, dtype: int64

Возраст клиента указан в диапазоне от 19 до 75 лет. Значение `0`, очевидно, является аномальным. Однако выяснить, каким оно должно было быть, уже не представляется возможным: это могло быть и `20`, и `30`, и `40`, и `50`, и `60`, и `70`.

Так как доля этих аномальных значений около 0,5%, удалим значение `0`, поскольку это практически не повлияет на итоговый результат.

In [14]:
# удаление нулей в возрасте
df = df.loc[df['age_years'] != 0]

Проверим полученный результат.

In [15]:
# вывод уникальных значений столбца 'age_years' и их частоты
df['age_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: age_years, dtype: int64

Теперь можем провести замену аномальных значений стажа на максимально возможное значение с учетом возраста каждого клиента.

In [16]:
# замена аномальных значений стажа на максимально возможное значение с учетом возраста каждого клиента
df.loc[df['days_employed'] > 22265, 'days_employed'] = 365 * (df['age_years']-14)

In [17]:
df.sort_values(by='days_employed', ascending = False)

children  days_employed  age_years education  education_id  \
4895          0        21900.0         74    высшее             0   
3460          0        21900.0         74   среднее             1   
2557          0        21900.0         74   среднее             1   
19642         0        21900.0         74   среднее             1   
229           0        21535.0         73   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

          family_status  family_status_id gender employment_type  debt  \
4895    женат / замужем                 0      F       пенсионер     0   
3460    женат / замужем                 0      M       пенсионер     0   
2557    женат / замужем                 0      F       пенсионер     0   
19642    вдовец / вдова                 2      F       пенсионер     0   
229      вдовец / вдова                 2      F       пенсионер     0   
...                 ...               ...    ...             ...   ...   
21489   женат / замужем                 0      M       компаньон     0   
21495  гражданский брак                 1      F       сотрудник     0   
21497   женат / замужем                 0      F       компаньон     0   
21502   женат / замужем                 0      F       сотрудник     0   
21510   женат / замужем                 0      F       сотрудник     0   

        total_income                           purpose  
4895   134935.354225              покупка своего жилья  
3460    54754.745517   операции со своей недвижимостью  
2557    42927.300898                        автомобили  
19642   45089.546419           приобретение автомобиля  
229    136897.008375  сделка с подержанным автомобилем  
...              ...                               ...  
21489            NaN              сделка с автомобилем  
21495            NaN                           свадьба  
21497            NaN        строительство недвижимости  
21502            NaN  строительство жилой недвижимости  
21510            NaN           приобретение автомобиля  

[21424 rows x 12 columns]

Рассмотрим уникальные значения столбца `education_id` и их частоту.

In [18]:
# вывод уникальных значений столбца 'education_id' и их частоты
df['education_id'].value_counts()

1    15169
0     5225
2      742
3      282
4        6
Name: education_id, dtype: int64

Выведем на экран первые 10 строк, охватывающие значения `3` и `4`.

In [19]:
# вывод первых 10 строк с максимальными значениями в столбце 'education_id'
df.sort_values(by='education_id', ascending = False).head(10)

children  days_employed  age_years       education  education_id  \
12021         3    5968.075884         36  ученая степень             4   
21519         1    2351.431934         37  ученая степень             4   
6551          0    5352.038180         58  ученая степень             4   
4170          0     409.200149         45  УЧЕНАЯ СТЕПЕНЬ             4   
2963          0   20075.000000         69  Ученая степень             4   
12786         0   17520.000000         62  ученая степень             4   
5263          3     805.239416         30       начальное             3   
12052         1   14965.000000         55       начальное             3   
9851          0     839.916738         45       начальное             3   
14356         0   17520.000000         62       начальное             3   

               family_status  family_status_id gender employment_type  debt  \
12021        женат / замужем                 0      F     госслужащий     0   
21519              в разводе                 3      M       сотрудник     0   
6551         женат / замужем                 0      M       сотрудник     0   
4170   Не женат / не замужем                 4      M       сотрудник     0   
2963         женат / замужем                 0      M       пенсионер     0   
12786        женат / замужем                 0      F       пенсионер     0   
5263               в разводе                 3      F       сотрудник     0   
12052       гражданский брак                 1      F       пенсионер     0   
9851         женат / замужем                 0      M       компаньон     1   
14356        женат / замужем                 0      F       пенсионер     1   

        total_income                                purpose  
12021  111392.231107                          покупка жилья  
21519  115949.039788      покупка коммерческой недвижимости  
6551   268411.214536           заняться высшим образованием  
4170   198570.757322  операции с коммерческой недвижимостью  
2963    98752.495442             покупка жилой недвижимости  
12786  255425.196556             покупка жилой недвижимости  
5263   122164.630609             строительство недвижимости  
12052   25308.586849             дополнительное образование  
9851   229539.360436               операции с недвижимостью  
14356  284301.643893                             автомобиль

Посмотрим на строки, в которых `education_id` равен 2.

In [20]:
# вывод строк с идентификатором '2' в столбце 'education_id'
df.loc[df['education_id'] == 2]

children  days_employed  age_years            education  education_id  \
13            0    1846.641941         54  неоконченное высшее             2   
42            0    1257.496190         20  неоконченное высшее             2   
43            0    4375.681384         43  неоконченное высшее             2   
64            0     118.891411         35  неоконченное высшее             2   
134           1    4171.107677         46  НЕОКОНЧЕННОЕ ВЫСШЕЕ             2   
...         ...            ...        ...                  ...           ...   
21432         1            NaN         38  неоконченное высшее             2   
21437         0     305.424394         25  неоконченное высшее             2   
21481         0     829.519627         30  неоконченное высшее             2   
21500         0     578.082757         26  неоконченное высшее             2   
21514         0     280.469996         27  неоконченное высшее             2   

               family_status  family_status_id gender employment_type  debt  \
13           женат / замужем                 0      F       сотрудник     0   
42     Не женат / не замужем                 4      F       сотрудник     0   
43                 в разводе                 3      F       компаньон     0   
64     Не женат / не замужем                 4      M       сотрудник     0   
134                в разводе                 3      F       сотрудник     0   
...                      ...               ...    ...             ...   ...   
21432  Не женат / не замужем                 4      F       сотрудник     0   
21437  Не женат / не замужем                 4      F       сотрудник     0   
21481       гражданский брак                 1      M       сотрудник     0   
21500  Не женат / не замужем                 4      M       компаньон     0   
21514  Не женат / не замужем                 4      M       компаньон     0   

        total_income                                purpose  
13     130458.228857                приобретение автомобиля  
42      82065.089021                  получение образования  
43     272957.133509                приобретение автомобиля  
64     282900.471308       сделка с подержанным автомобилем  
134    166015.790504             покупка жилой недвижимости  
...              ...                                    ...  
21432            NaN                      операции с жильем  
21437  147216.431172                          покупка жилья  
21481  440470.880955                  на покупку автомобиля  
21500   77813.293365  операции с коммерческой недвижимостью  
21514  355988.407188             строительство недвижимости  

[742 rows x 12 columns]

Получается нелогичная градация: 0 - высшее образование, 1 - среднее образование, 2 - незаконченное высшее, 3 - начальное образование, а 4 - ученая степень.

Правильнее будет присвоить id следующим образом: 0 - начальное образование, 1 - среднее образование, 2 - незаконченное высшее образование, 3 - высшее образование, 4 - ученая степень. Необходимо заменить id `0` на `3`, и наоборот. Чтобы не смешать значения `0` и `3`, проведем замену в несколько итераций с промежуточным идентификатором `30`.

In [21]:
# замена нелогичных значений 'education_id'
df['education_id']=df['education_id'].replace([0], 30)
df['education_id']=df['education_id'].replace([3], 0)
df['education_id']=df['education_id'].replace([30], 3)

Проверим результат, выведя на экран значения с идентификатором `0`.

In [22]:
# вывод первых пяти строк с идентификатором '0' в столбце 'education_id'
df.loc[df['education_id'] == 0].head()

children  days_employed  age_years  education  education_id  \
31          0    1682.083438         47  начальное             0   
136         0   16790.000000         60  начальное             0   
210         2   14965.000000         55  начальное             0   
211         0     577.273434         41  начальное             0   
272         0    5538.963731         35  начальное             0   

        family_status  family_status_id gender employment_type  debt  \
31    женат / замужем                 0      F       сотрудник     0   
136   женат / замужем                 0      M       пенсионер     0   
210    вдовец / вдова                 2      F       пенсионер     0   
211  гражданский брак                 1      F       сотрудник     0   
272  гражданский брак                 1      M       сотрудник     1   

      total_income                           purpose  
31   275485.684538                             жилье  
136  113124.202781   операции со своей недвижимостью  
210  160635.120347  строительство жилой недвижимости  
211   36649.078439  сделка с подержанным автомобилем  
272  123814.079468             заняться образованием

Всё нормально.

Рассмотрим уникальные значения столбца `family_status_id` и их частоту.

In [23]:
# вывод уникальных значений столбца 'family_status_id' и их частоты
df['family_status_id'].value_counts()

0    12331
1     4156
4     2797
3     1185
2      955
Name: family_status_id, dtype: int64

Выведем последовательно на экран строки, охватывающие значения `2`, `3` и `4`.

In [24]:
# вывод первых пяти строк с идентификатором '2' в столбце 'family_status_id'
df.loc[df['family_status_id'] == 2].head()

children  days_employed  age_years education  education_id  \
18          0   14235.000000         53   среднее             1   
56          0   18250.000000         64   среднее             1   
57          0    2424.809749         44   среднее             1   
100         0     605.026441         49   СРЕДНЕЕ             1   
105         0    2098.626296         62   СРЕДНЕЕ             1   

      family_status  family_status_id gender employment_type  debt  \
18   вдовец / вдова                 2      F       пенсионер     0   
56   вдовец / вдова                 2      F       пенсионер     0   
57   вдовец / вдова                 2      F       сотрудник     0   
100  вдовец / вдова                 2      F       сотрудник     1   
105  вдовец / вдова                 2      F       сотрудник     0   

      total_income                                 purpose  
18    56823.777243      на покупку подержанного автомобиля  
56   149141.043533                             образование  
57    58616.405960                             образование  
100  176384.882536                операции с недвижимостью  
105   76884.187588  строительство собственной недвижимости

In [25]:
# вывод первых пяти строк с идентификатором '3' в столбце 'family_status_id'
df.loc[df['family_status_id'] == 3].head()

children  days_employed  age_years            education  education_id  \
19          0   10038.818549         48              СРЕДНЕЕ             1   
43          0    4375.681384         43  неоконченное высшее             2   
53          0    4219.883386         61              СРЕДНЕЕ             1   
117         0    3817.113295         41              среднее             1   
128         0   19345.000000         67              среднее             1   

    family_status  family_status_id gender employment_type  debt  \
19      в разводе                 3      F       сотрудник     0   
43      в разводе                 3      F       компаньон     0   
53      в разводе                 3      F       компаньон     0   
117     в разводе                 3      F       сотрудник     0   
128     в разводе                 3      F       пенсионер     0   

      total_income                       purpose  
19   242831.107982  на покупку своего автомобиля  
43   272957.133509       приобретение автомобиля  
53   311453.599530          покупка своего жилья  
117  178975.133574         на покупку автомобиля  
128  112331.787067          сделка с автомобилем

In [26]:
# вывод первых пяти строк с идентификатором '4' в столбце 'family_status_id'
df.loc[df['family_status_id'] == 4].head()

children  days_employed  age_years education  education_id  \
24         1   15695.000000         57   среднее             1   
29         0            NaN         63   среднее             1   
36         0     176.216688         33   среднее             1   
38         0     597.881827         25    высшее             3   
40         0    1030.362606         30   среднее             1   

            family_status  family_status_id gender employment_type  debt  \
24  Не женат / не замужем                 4      F       пенсионер     0   
29  Не женат / не замужем                 4      F       пенсионер     0   
36  Не женат / не замужем                 4      M       сотрудник     0   
38  Не женат / не замужем                 4      M       сотрудник     1   
40  Не женат / не замужем                 4      M       компаньон     0   

     total_income                                purpose  
24  290547.235997  операции с коммерческой недвижимостью  
29            NaN       строительство жилой недвижимости  
36  138830.650734                             автомобили  
38  192247.298639                            образование  
40  142865.579146                   покупка недвижимости

В данном случае идентификаторы можно поменять местами (0 - не женат / не замужем, 1 - гражданский брак, 2 - женат / замужем, 3 - в разводе, 4 - вдовец / вдова), а можно и не менять. Оставим как есть.

Проверим также, есть ли аномальные значения в столбце `gender`.

In [27]:
# вывод уникальных значений столбца 'gender' и их частоты
df['gender'].value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

Одно из значений в столбце `gender`, очевидно, неправильное. Исправить это значение невозможно, а поскольку это одна строчка, её можно удалить.

In [28]:
# исключение из таблицы df строк, в которых значение столбца 'gender' равно 'XNA'
df = df.loc[df['gender'] != 'XNA']

Проверим столбец `gender`.

In [29]:
# вывод уникальных значений столбца 'gender' и их частоты
df['gender'].value_counts()

F    14164
M     7259
Name: gender, dtype: int64

Мы заменили все аномальные значения, пойдём дальше.

### Пропуски значений
Посчитаем, сколько в таблице пропущенных значений. Для этого достаточно двух методов `pandas`:

In [30]:
# подсчёт пропусков
df.isna().sum()

children               0
days_employed       2164
age_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
employment_type        0
debt                   0
total_income        2164
purpose                0
dtype: int64

Все пропуски находятся только в двух столбцах — `days_employed` и `total_income`. Однако их достаточно много (10% данных) и они важны для цели исследования, поэтому заменить их на нули нельзя. Посмотрим, в каких строках присутствуют пропуски и не вызваны ли они чем-нибудь. Выведем на экран 25 строчек с пропущенными данными в столбце `days_employed`.

In [31]:
# получение первых 25 строк таблицы с пропущенными значениями в столбце days_employed
df[df['days_employed'].isna()].head(25)

children  days_employed  age_years            education  education_id  \
12          0            NaN         65              среднее             1   
26          0            NaN         41              среднее             1   
29          0            NaN         63              среднее             1   
41          0            NaN         50              среднее             1   
55          0            NaN         54              среднее             1   
65          0            NaN         21              среднее             1   
67          0            NaN         52               высшее             3   
72          1            NaN         32               высшее             3   
82          2            NaN         50               высшее             3   
83          0            NaN         52              среднее             1   
90          2            NaN         35               высшее             3   
94          1            NaN         34               высшее             3   
96          0            NaN         44              СРЕДНЕЕ             1   
97          0            NaN         47               высшее             3   
120         0            NaN         46              среднее             1   
121         0            NaN         29               высшее             3   
135         0            NaN         27              среднее             1   
141         0            NaN         39              среднее             1   
145         0            NaN         62              среднее             1   
174         0            NaN         55               высшее             3   
181         0            NaN         26              среднее             1   
189         1            NaN         30              среднее             1   
205         1            NaN         31               высшее             3   
220         1            NaN         23  неоконченное высшее             2   
241         0            NaN         47              среднее             1   

             family_status  family_status_id gender employment_type  debt  \
12        гражданский брак                 1      M       пенсионер     0   
26         женат / замужем                 0      M     госслужащий     0   
29   Не женат / не замужем                 4      F       пенсионер     0   
41         женат / замужем                 0      F     госслужащий     0   
55        гражданский брак                 1      F       пенсионер     1   
65   Не женат / не замужем                 4      M       компаньон     0   
67         женат / замужем                 0      F       пенсионер     0   
72         женат / замужем                 0      M     госслужащий     0   
82         женат / замужем                 0      F       сотрудник     0   
83         женат / замужем                 0      M       сотрудник     0   
90         женат / замужем                 0      F       сотрудник     0   
94        гражданский брак                 1      F       компаньон     0   
96         женат / замужем                 0      F       сотрудник     0   
97         женат / замужем                 0      F       сотрудник     0   
120        женат / замужем                 0      F       сотрудник     0   
121        женат / замужем                 0      F       компаньон     0   
135        женат / замужем                 0      M       компаньон     0   
141       гражданский брак                 1      M       сотрудник     0   
145        женат / замужем                 0      M       пенсионер     0   
174         вдовец / вдова                 2      F       компаньон     0   
181       гражданский брак                 1      F       компаньон     1   
189  Не женат / не замужем                 4      F       сотрудник     0   
205        женат / замужем                 0      F       сотрудник     0   
220       гражданский брак                 1      F       компаньон     0   
241        женат / замужем                 0      M

Мы видим, что пропуски в столбце `days_employed` нашей выборки соответствуют пропускам в столбце `total_income`. При этом никакой связи этих пропусков с другими столбцами нет. Возможно, пропуски связаны с нежеланием людей указывать свой трудовой стаж и доход, либо нежеланием заниматься подсчетами. А, возможно, банковские менеджеры не стали заполнять данные колонки при отсутствии документов, подтверждающих стаж и доход.

Чтобы не исказить имеющиеся взаимосвязи, заменим пропущенные значения в столбцах `days_employed` и `total_income` медианными значениями. Выбираем последние в качестве замены, т.к. медиана - это значение, которое делит количественную категорию пополам.

In [32]:
# замена пропущенных значений на медиану
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())

Убедимся, что в таблице не осталось пропусков. Для этого ещё раз посчитаем пропущенные значения.

In [33]:
# подсчёт пропусков
df.isna().sum()

children            0
days_employed       0
age_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
employment_type     0
debt                0
total_income        0
purpose             0
dtype: int64

### Дубликаты
Теперь обратимся к дубликатам. Посчитаем явные дубликаты в таблице одной командой:

In [34]:
# подсчёт явных дубликатов
df.duplicated().sum()

54

Вызовем специальный метод `pandas`, чтобы удалить явные дубликаты:

In [35]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop = True)

Ещё раз посчитаем явные дубликаты в таблице — убедимся, что полностью от них избавились:

In [36]:
# проверка на отсутствие дубликатов
df.duplicated().sum()

0

Теперь избавимся от неявных дубликатов. 
Как мы помним, неявные дубликаты были обнаружены в колонке `children`: значения `20` вместо `2`. 
В колонке `education` проставлены значения с разным регистром. Их также нужно привести к единому формату.

Начнём с количества детей.

In [37]:
# замена неявных дубликатов в столбце 'children'
df['children']=df['children'].replace([20], 2)

Проверим полученный результат.

In [38]:
# оценка числовых переменных
df.describe()

children  days_employed     age_years  education_id  \
count  21369.000000   21369.000000  21369.000000  21369.000000   
mean       0.480182    4620.065326     43.484534      1.510506   
std        0.755668    5667.486652     12.245590      0.875321   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000    1024.308432     33.000000      1.000000   
50%        0.000000    2197.320350     43.000000      1.000000   
75%        1.000000    4791.848816     53.000000      2.000000   
max        5.000000   21900.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21369.000000  21369.000000  2.136900e+04  
mean           0.972437      0.081099  1.652458e+05  
std            1.420533      0.272993  9.809662e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077064e+05  
50%            0.000000      0.000000  1.450055e+05  
75%            1.000000      0.000000  1.957379e+05  
max            4.000000      1.000000  2.265604e+06

Теперь обратимся к колонке `education`. Приведём все символы в ней к нижнему регистру:

In [39]:
# приведение всех символов в колонке к нижнему регистру
df['education'] = df['education'].str.lower()

Для проверки полученного результата выведем на экран список уникальных значений в столбце, отсортированный в алфавитном порядке.

In [40]:
# Просмотр уникальных значений столбца 'education'
sorted(df['education'].unique())

['высшее', 'начальное', 'неоконченное высшее', 'среднее', 'ученая степень']

Теперь всё в порядке.

Проверим, есть ли неявные дубликаты в столбцах `family_status` и `employment_type`.

In [41]:
# Проверка на неявные дубликаты столбца 'family_status'
sorted(df['family_status'].unique())

['Не женат / не замужем',
 'в разводе',
 'вдовец / вдова',
 'гражданский брак',
 'женат / замужем']

In [42]:
# Проверка на неявные дубликаты столбца 'employment_type'
sorted(df['employment_type'].unique())

['безработный',
 'в декрете',
 'госслужащий',
 'компаньон',
 'пенсионер',
 'предприниматель',
 'сотрудник',
 'студент']

Всё в порядке.

Неявные дубликаты остались в столбце `purpose`. Однако вместо их исправления нам потребуется объединить цели кредита в обобщённые категории, поэтому оставим как есть.

Получим заново общую информацию о нашей таблице после всех трансформаций:

In [43]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21369 entries, 0 to 21368
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21369 non-null  int64  
 1   days_employed     21369 non-null  float64
 2   age_years         21369 non-null  int64  
 3   education         21369 non-null  object 
 4   education_id      21369 non-null  int64  
 5   family_status     21369 non-null  object 
 6   family_status_id  21369 non-null  int64  
 7   gender            21369 non-null  object 
 8   employment_type   21369 non-null  object 
 9   debt              21369 non-null  int64  
 10  total_income      21369 non-null  float64
 11  purpose           21369 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Категоризация данных
Проведём заключительную трансформацию данных. 

Вначале заменим все вещественные данные на целочисленные. Это нужно будет сделать для столбцов `days_employed` и `total_income`.

In [44]:
# замена вещественных данных на целочисленные с помощью метода astype()
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

Проверим полученный результат, выведя на экран первые 10 строк таблицы.

In [45]:
# получение первых 10 строк таблицы df
df.head(10)

children  days_employed  age_years education  education_id  \
0         1           8437         42    высшее             3   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14235         53   среднее             1   
5         0            926         27    высшее             3   
6         0           2879         43    высшее             3   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             3   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender employment_type  debt  \
0   женат / замужем                 0      F       сотрудник     0   
1   женат / замужем                 0      F       сотрудник     0   
2   женат / замужем                 0      M       сотрудник     0   
3   женат / замужем                 0      M       сотрудник     0   
4  гражданский брак                 1      F       пенсионер     0   
5  гражданский брак                 1      M       компаньон     0   
6   женат / замужем                 0      F       компаньон     0   
7   женат / замужем                 0      M       сотрудник     0   
8  гражданский брак                 1      F       сотрудник     0   
9   женат / замужем                 0      M       сотрудник     0   

   total_income                     purpose  
0        253875               покупка жилья  
1        112080     приобретение автомобиля  
2        145885               покупка жилья  
3        267628  дополнительное образование  
4        158616             сыграть свадьбу  
5        255763               покупка жилья  
6        240525           операции с жильем  
7        135823                 образование  
8         95856       на проведение свадьбы  
9        144425     покупка жилья для семьи

Получилось.

Теперь создадим для удобства два новых датафрейма-"словаря" `education_dict` и `family_status_dict`, включив туда идентификаторы уровня образования и семейного статуса.

In [46]:
# создание двух новых таблиц-"словарей"
education_dict = df[['education', 'education_id']]
family_status_dict = df[['family_status', 'family_status_id']]

Удалим из исходной таблицы колонки `education` и `family_status`, оставив только идентификаторы.

In [47]:
# удаление колонок 'education' и 'family_status'
df = df[['children', 'days_employed', 'age_years', 'education_id', 'family_status_id', 'gender', 'employment_type', 'debt',\
         'total_income', 'purpose']]

Выведем на экран полученные "словари", сгруппировав клиентов по образованию и семейному положению.

In [48]:
# получение распределения клиентов в зависимости от идентификатора образования в "словаре" 'education_dict'
education_dict.groupby('education_id').count()

education
education_id           
0                   282
1                 15124
2                   741
3                  5216
4                     6

In [49]:
# получение распределения клиентов в зависимости от идентификатора семейного положения в "словаре" 'family_status_dict'
family_status_dict.groupby('family_status_id').count()

family_status
family_status_id               
0                         12295
1                          4141
2                           954
3                          1185
4                          2794

In [50]:
# получение первых 10 строк таблицы df
df.head()

children  days_employed  age_years  education_id  family_status_id gender  \
0         1           8437         42             3                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          14235         53             1                 1      F   

  employment_type  debt  total_income                     purpose  
0       сотрудник     0        253875               покупка жилья  
1       сотрудник     0        112080     приобретение автомобиля  
2       сотрудник     0        145885               покупка жилья  
3       сотрудник     0        267628  дополнительное образование  
4       пенсионер     0        158616             сыграть свадьбу

Разделим клиентов на категории в зависимости от уровня доходов:

* 0 – 30000 — `'E'`;
* 30001 – 50000 — `'D'`;
* 50001 – 200000 — `'C'`;
* 200001 – 1000000 — `'B'`;
* 1000001 и выше — `'A'`.

Укажем эти категории в новом столбце `total_income_category`.

In [51]:
# создадим функцию, разделяющую клиентов на категории в зависимости от уровня дохода
def income_category(row):
    total_income = row['total_income']
    if total_income <= 30000:
        return 'E'
    if total_income >= 30001 and total_income <= 50000:
        return 'D'
    if total_income >= 50001 and total_income <= 200000:
        return 'C'
    if total_income >= 200001 and total_income <= 1000000:
        return 'B'
    if total_income >= 1000001:
        return 'A'
    return income_category

df['total_income_category'] = df.apply(income_category, axis = 1)

Выведем на экран полученную таблицу.

In [52]:
# получение первых 10 строк таблицы df
df.head(10)

children  days_employed  age_years  education_id  family_status_id gender  \
0         1           8437         42             3                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          14235         53             1                 1      F   
5         0            926         27             3                 1      M   
6         0           2879         43             3                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             3                 1      F   
9         0           2188         41             1                 0      M   

  employment_type  debt  total_income                     purpose  \
0       сотрудник     0        253875               покупка жилья   
1       сотрудник     0        112080     приобретение автомобиля   
2       сотрудник     0        145885               покупка жилья   
3       сотрудник     0        267628  дополнительное образование   
4       пенсионер     0        158616             сыграть свадьбу   
5       компаньон     0        255763               покупка жилья   
6       компаньон     0        240525           операции с жильем   
7       сотрудник     0        135823                 образование   
8       сотрудник     0         95856       на проведение свадьбы   
9       сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

Сгруппируем цели кредита в новом столбце `purpose_category`. Для этого посмотрим, какие есть уникальные значения в столбце `purpose`.

In [53]:
# получение уникальных значений колонки 'purpose'
sorted(df['purpose'].unique())

['автомобили',
 'автомобиль',
 'высшее образование',
 'дополнительное образование',
 'жилье',
 'заняться высшим образованием',
 'заняться образованием',
 'на покупку автомобиля',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'на проведение свадьбы',
 'недвижимость',
 'образование',
 'операции с жильем',
 'операции с коммерческой недвижимостью',
 'операции с недвижимостью',
 'операции со своей недвижимостью',
 'покупка жилой недвижимости',
 'покупка жилья',
 'покупка жилья для сдачи',
 'покупка жилья для семьи',
 'покупка коммерческой недвижимости',
 'покупка недвижимости',
 'покупка своего жилья',
 'получение высшего образования',
 'получение дополнительного образования',
 'получение образования',
 'приобретение автомобиля',
 'профильное образование',
 'ремонт жилью',
 'свадьба',
 'свой автомобиль',
 'сделка с автомобилем',
 'сделка с подержанным автомобилем',
 'строительство жилой недвижимости',
 'строительство недвижимости',
 'строительство собственной недв

Объединим все цели в следующие категории `purpose_category`:
    
* 'операции с автомобилем',
* 'операции с недвижимостью',
* 'проведение свадьбы',
* 'получение образования'.

Создадим функцию для заполнения столбца `purpose_category`.

In [54]:
# Функция для заполнения категорий в столбце `purpose_category`
def purpose_category(row):
    purpose = row['purpose']
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'свад' in purpose:
        return 'проведение свадьбы'
    if 'образ' in purpose:
        return 'получение образования'
    return 'операции с недвижимостью'
df['purpose_category'] = df.apply(purpose_category, axis = 1)

In [55]:
# получение первых 10 строк таблицы df
df.head(10) 

children  days_employed  age_years  education_id  family_status_id gender  \
0         1           8437         42             3                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          14235         53             1                 1      F   
5         0            926         27             3                 1      M   
6         0           2879         43             3                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             3                 1      F   
9         0           2188         41             1                 0      M   

  employment_type  debt  total_income                     purpose  \
0       сотрудник     0        253875               покупка жилья   
1       сотрудник     0        112080     приобретение автомобиля   
2       сотрудник     0        145885               покупка жилья   
3       сотрудник     0        267628  дополнительное образование   
4       пенсионер     0        158616             сыграть свадьбу   
5       компаньон     0        255763               покупка жилья   
6       компаньон     0        240525           операции с жильем   
7       сотрудник     0        135823                 образование   
8       сотрудник     0         95856       на проведение свадьбы   
9       сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

Оценим ещё раз числовые данные.

In [56]:
# оценка числовых переменных
df.describe()

children  days_employed     age_years  education_id  \
count  21369.000000   21369.000000  21369.000000  21369.000000   
mean       0.480182    4619.664046     43.484534      1.510506   
std        0.755668    5667.649078     12.245590      0.875321   
min        0.000000      24.000000     19.000000      0.000000   
25%        0.000000    1024.000000     33.000000      1.000000   
50%        0.000000    2197.000000     43.000000      1.000000   
75%        1.000000    4791.000000     53.000000      2.000000   
max        5.000000   21900.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21369.000000  21369.000000  2.136900e+04  
mean           0.972437      0.081099  1.652453e+05  
std            1.420533      0.272993  9.809662e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.077060e+05  
50%            0.000000      0.000000  1.450050e+05  
75%            1.000000      0.000000  1.957370e+05  
max            4.000000      1.000000  2.265604e+06

**Выводы**

Предобработка обнаружила целый ряд критичных и некритичных проблем:

- нарушения в стиле заголовков,
- аномальные значения (минусовые и слишком большие значения, несуществующий пол, нелогичные идентификаторы),
- пропущенные значения,
- дубликаты — явные и неявные.

Мы исправили названия двух столбцов и заменили 2 164 пропущенных значения в столбцах `days_employed` и `total_income` на медианные значения.

Также удалили 55 явных дубликатов записей, 101 ноль в столбце возраста, изменили 76 неявных дубликатов и удалили одно аномальное значение пола.
В итоге удаление затронуло всего 0,73% записей - это не повлияет на точность нашего исследования. Оставшихся 21 369 уникальных ненулевых записей достаточно для того, чтобы сделать релевантные выводы о поведении клиентов.

Кроме того, мы провели категоризацию данных и создали две таблицы-"словаря", разгрузив основную таблицу.

В среднем, 8,1% клиентов допускают просрочку кредита. Попробуем выяснить, от каких факторов зависит этот показатель.





## Определение факторов, влияющих на вероятность просрочки кредита

### Влияние количества детей на возврат кредита

Первый вопрос нашего заказчика касался влияния количества детей на вероятность возврата кредита в срок.

Для этого:

* Создадим сводную таблицу `pivot_children`, в которой сгруппируем данные о просрочке кредита для клиентов с различным числом детей.
* Сравним, каков процент выхода на просрочку кредита в каждой группе.


Посчитаем общее количество клиентов в каждой группе и количество клиентов, допустивших просрочку кредита.

In [57]:
# создание сводной таблицы 'pivot_children', в которой группируем данные по столбцу 'children'
# и подсчитываем для этих групп количество клиентов 'count' и количество просрочек 'sum'
pivot_children = df.pivot_table(index=['children'], values='debt', aggfunc=['sum', 'count'])
pivot_children

sum  count
          debt   debt
children             
0         1058  14037
1          442   4840
2          202   2114
3           27    328
4            4     41
5            0      9

Посчитаем процент просрочки кредита в каждой группе. Оказывается, в группе с пятью детьми просрочек не было.

In [58]:
# добавление в сводную таблицу 'pivot_children' столбца с процентом просрочки
pivot_children['default'] = 100 * (pivot_children['sum'] / pivot_children['count'])
pivot_children

sum  count   default
          debt   debt          
children                       
0         1058  14037  7.537223
1          442   4840  9.132231
2          202   2114  9.555345
3           27    328  8.231707
4            4     41  9.756098
5            0      9  0.000000

**Выводы**

Данные не показывают связи между количеством детей и выходом на просрочку кредита:

- Пик просрочки приходится на клиентов с одним, двумя и четырьмя детьми. Клиенты с трёмя детьми показывают среднюю вероятность просрочки кредита.
- Клиенты без детей выходят на просрочку кредита значительно реже, чем семьи с детьми.
- Клиенты с пятью детьми вообще не допускали просрочки кредита. Однако количество таких клиентов слишком мало (9).

**Таким образом, на вероятность просрочки кредита влияет не количество детей, а их наличие или отсутствие.**

Всё логично: появление даже одного ребёнка приводит к дополнительным расходам семьи без увеличения (а поначалу и с уменьшением) доходов, что делает более напряжённым выполнение графика погашения задолженности перед банком. Отсутствие просрочки у клиентов с пятью детьми, возможно, связано с различными государственными программами поддержки таких семей либо просто с маленьким числом таких клиентов, а следовательно, с нерепрезентативностью выборки в данной категории.

### Влияние семейного положения на возврат кредита

Согласно второй задаче, необходимо выяснить, влияет ли семейное положение на вероятность возникновения просрочки по кредиту.

Чтобы ответить на этот вопрос, проделаем действия, аналогичные предыдущему параграфу.

Для этого:

* Создадим сводную таблицу `pivot_family`, в которой сгруппируем данные о просрочке кредита для клиентов с разным семейным положением.
* Сравним, каков процент выхода на просрочку кредита в каждой группе.


Посчитаем общее количество клиентов в каждой группе и количество клиентов, допустивших просрочку кредита.

In [59]:
# создание сводной таблицы 'pivot_family', в которой группируем данные по столбцу 'family_status_id'
# и подсчитываем для этих групп количество клиентов 'count' и количество просрочек 'sum'
pivot_family = df.pivot_table(index=['family_status_id'], values=['debt'], aggfunc=['sum', 'count'])
pivot_family

sum  count
                 debt   debt
family_status_id            
0                 927  12295
1                 386   4141
2                  62    954
3                  85   1185
4                 273   2794

Посчитаем процент просрочки кредита в каждой группе.

In [60]:
# добавление в сводную таблицу 'pivot_family' столбца с процентом просрочки
pivot_family['default'] = 100 * (pivot_family['sum'] / pivot_family['count'])
pivot_family

sum  count   default
                 debt   debt          
family_status_id                      
0                 927  12295  7.539650
1                 386   4141  9.321420
2                  62    954  6.498952
3                  85   1185  7.172996
4                 273   2794  9.770938

**Выводы**

Данные показывают взаимосвязь между семейным положением и выходом на просрочку кредита:

- Пик просрочки приходится на клиентов, которые ни разу не брали на себя семейные обязательства (клиенты со статусом "не женат / не замужем" и "гражданский брак").
- У остальных категорий клиентов вероятность просрочки по кредиту ниже среднего (8,1%).
- Минимальная вероятность просрочки у вдов и вдовцов (6,6%). Хотя число таких клиентов и невысоко (959), однако статистически значимо (4,5% выборки).

**Таким образом, семейное положение влияет на вероятность просрочки по кредиту.**

Всё логично: человек, хотя бы раз взявший на себя семейные обязательства, более ответственно подходит и к взятию на себя других обязательств, и к их выполнению.

In [65]:
#строим сводную таблицу
family_pivot = df.pivot_table(index='family_status_id', values='debt').reset_index()
family_pivot

family_status_id      debt
0                 0  0.075397
1                 1  0.093214
2                 2  0.064990
3                 3  0.071730
4                 4  0.097709

In [71]:
#удаляем дубликаты из словаря
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

In [73]:
#соединяем)
family_pivot.merge(family_status_dict, on='family_status_id').sort_values(by='debt', ascending=True)

family_status_id      debt          family_status
2                 2  0.064990         вдовец / вдова
3                 3  0.071730              в разводе
0                 0  0.075397        женат / замужем
1                 1  0.093214       гражданский брак
4                 4  0.097709  Не женат / не замужем

### Влияние уровня дохода на возврат кредита

Теперь выясним, влияет ли на вероятность возникновения просрочки по кредиту уровень дохода клиента.

Чтобы ответить на этот вопрос, проделаем действия, аналогичные предыдущему параграфу.

Для этого:

* Создадим сводную таблицу `pivot_income`, в которой сгруппируем данные о просрочке кредита для клиентов с разным уровнем дохода.
* Сравним, каков процент выхода на просрочку кредита в каждой группе.


Посчитаем общее количество клиентов в каждой группе и количество клиентов, допустивших просрочку кредита.

In [61]:
# создание сводной таблицы 'pivot_income', в которой группируем данные по столбцу 'total_income_category'
# и подсчитываем для этих групп количество клиентов 'count' и количество просрочек 'sum'
pivot_income = df.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum', 'count'])
pivot_income

sum  count
                       debt   debt
total_income_category             
A                         2     25
B                       355   5014
C                      1353  15960
D                        21    348
E                         2     22

Посчитаем процент просрочки кредита в каждой группе.

In [62]:
# добавление в сводную таблицу 'pivot_income' столбца с процентом просрочки
pivot_income['default'] = 100 * (pivot_income['sum'] / pivot_income['count'])
pivot_income

sum  count   default
                       debt   debt          
total_income_category                       
A                         2     25  8.000000
B                       355   5014  7.080176
C                      1353  15960  8.477444
D                        21    348  6.034483
E                         2     22  9.090909

**Выводы**

Данные показывают неоднозначную связь между уровнем дохода и выходом на просрочку кредита:

- Пик просрочки приходится на клиентов, имеющих минимальный доход (менее 30000) — 9,1%. Однако их число невелико для достоверных выводов.
- Клиенты со средним доходом (категория "С") также имеют вероятность просрочки выше среднего (8,48%).
- Минимальная вероятность просрочки у клиентов категории "D" (6%).

**Таким образом, уровень дохода скорее не влияет на вероятность просрочки по кредиту.**

Всё логично только для категории с самым маленьким доходом: человек, едва сводящий концы с концами и не имеющий сбережений, более подвержен выходу на просрочку.

### Влияние цели кредита на его возврат в срок

Наконец, оценим влияние цели кредита на вероятность возникновения просрочки по нему.

Создадим сводную таблицу `pivot_purpose`, в которой сгруппируем данные о просрочке кредита для клиентов с разными целями кредитования.
Сравним, каков процент выхода на просрочку кредита в каждой группе.


Посчитаем общее количество клиентов в каждой группе и количество клиентов, допустивших просрочку кредита.

In [63]:
# создание сводной таблицы 'pivot_purpose', в которой группируем данные по столбцу 'purpose_category'
# и подсчитываем для этих групп количество клиентов 'count' и количество просрочек 'sum'
pivot_purpose = df.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum', 'count'])
pivot_purpose

sum  count
                         debt   debt
purpose_category                    
операции с автомобилем    400   4286
операции с недвижимостью  779  10766
получение образования     370   3996
проведение свадьбы        184   2321

Посчитаем процент просрочки кредита в каждой группе.

In [64]:
# добавление в сводную таблицу 'pivot_purpose' столбца с процентом просрочки
pivot_purpose['default'] = 100 * (pivot_purpose['sum'] / pivot_purpose['count'])
pivot_purpose

sum  count   default
                         debt   debt          
purpose_category                              
операции с автомобилем    400   4286  9.332711
операции с недвижимостью  779  10766  7.235742
получение образования     370   3996  9.259259
проведение свадьбы        184   2321  7.927617

**Выводы**

Данные показывают статистически выраженную связь между целью кредита и вероятностью выхода на просрочку:

- Пик просрочки приходится на клиентов, берущих кредит на операции с автомобилем и на получение образования (9,3%).
- Кредиты на проведение свадьбы имеют вероятность просрочки, близкую к средней (7,9%).
- Минимальная вероятность просрочки у кредитов, взятых на операции с недвижимостью (7,2%).

**Таким образом, цель кредита влияет на вероятность просрочки по кредиту.**

Всё логично: человек, берущий кредит на операции с недвижимостью, приложит максимум усилий для выполнения своих обязательств перед банком, чтобы не остаться без жилья (которое в залоге). О влиянии свадьбы на вероятность просрочки по кредиту мы выяснили выше.

## Итоги исследования

Мы проверили четыре фактора и установили:

1. Само по себе количество детей не влияет на вероятность просрочки по кредиту. Вероятность повышается при наличии хотя бы одного ребёнка: клиенты без детей имеют меньше просрочек.


2. Семейное положение влияет на вероятность просрочки. Человек, хотя бы раз взявший на себя семейные обязательства, более ответственно подходит и к взятию на себя других обязательств, и к их выполнению.


3. Уровень дохода скорее не влияет на вероятность просрочки по кредиту. Исключение - клиенты с доходом до 30000.


4. Цель кредита влияет на вероятность просрочки: у кредитов, взятых на операции с недвижимостью, вероятность просрочки ниже, у кредитов на проведение свадьбы - на среднем уровне, у образовательных и автокредитов - вероятность просрочки выше среднего.
